helper/fairshare/jobmanager.go

```diff
@@ -179,10 +177,94 @@ func (j *JobManager) getNextJob() Job {
 	}
 
 	if j.queues[queueID].Len() == 0 {
-		j.removeLastQueueAccessed()
+		// we remove the empty queue, but we don't remove the worker count
+		// in case we are still working on previous jobs from this queue.
+		// worker count cleanup is handled in j.decrementWorkerCount
+		delete(j.queues, queueID)
+	}
+
+	return out.(Job), queueID
+}
+
+// returns the next queue to assign work from, and a bool if there is a queue
+// that can have a worker assigned.
+// the intent is to avoid over-allocating work from specific queues, as
+// outlined in RFC VLT-145
+// note: this must be called with j.l held
+func (j *JobManager) getNextQueue() (string, bool) {
+	var nextQueue string
+	var canAssignWorker bool
+
+	queueIDsByIncreasingWorkers := j.sortByNumWorkers()
+	for _, queueID := range queueIDsByIncreasingWorkers {
+		if !j.queueWorkersSaturated(queueID) {
+			nextQueue = queueID
+			canAssignWorker = true
+			break
+		}
 	}
 
-	return out.(Job)
+	return nextQueue, canAssignWorker
+}
+
+// returns true if there are already too many workers on this queue
+// note: this must be called with j.l held (at least for read).
+// note: we may want to eventually factor in queue length relative to num queues
+func (j *JobManager) queueWorkersSaturated(queueID string) bool {
+	numActiveQueues := float64(len(j.queues))
+	numTotalWorkers := float64(j.workerPool.numWorkers)
+	maxWorkersPerQueue := math.Ceil(0.9 * numTotalWorkers / numActiveQueues)
+
+	numWorkersPerQueue := j.workerCount
+
+	return numWorkersPerQueue[queueID] >= int(maxWorkersPerQueue)
+}
+
+// sortByNumWorkers returns queueIDs in order of increasing number of workers
+// note: this must be called with j.l held
+func (j *JobManager) sortByNumWorkers() []string {
+	out := make([]string, 0, len(j.queues))
+	for queueID := range j.queues {
+		out = append(out, queueID)
+	}
+
+	workersPerQueue := j.workerCount
+
+	sort.Slice(out, func(i, j int) bool {
+		// TODO should we be explicitly breaking times with the queueID, or
+		// might we want some randomness?
+		// I think it's probably fine since it only breaks ties between queues
+		// with equal workers assigned, and makes testing much easier
+		if workersPerQueue[out[i]] == workersPerQueue[out[j]] {
```

Comment:
I don't think you need this check. If you want the sorting to be stable, you could call `SliceStable` , which I think should allow you to remove this code. But, by virtue of this being a tiebreaker, I don't think we need stability.

Reply:
i read the doc for `Slice` and `SliceStable` and didn't see anything about them breaking ties in this way. do you know how they break ties?

Since this data comes from a map (which has no guaranteed ordering) i think we either accept the inherent randomness of that map's order (i believe the iter is random), or we explicitly sort.

My thought process is explicitly sorting is nice for the testing process so we don't have flakiness if the map gets unordered?

Reply:
Oh, I see, no that makes perfect sense! I didn't realize that the comment above this code was talking about the inherent randomness in the unstable sorting. Feel free to resolve this convo!